Allow multiple `root` routes in same scope level

rafaelsales opened the pull request:
TL;DR:
**Currently I have:**

``` ruby
get '/', to: 'portfolio#show', constraints: ->(req) { Hostname.portfolio_site?(req.host) }
get '/', to: 'blog#show',      constraints: ->(req) { Hostname.blog_site?(req.host) }
root 'landing#show'
```

**But I would like to have:**

``` ruby
root 'portfolio#show', constraints: ->(req) { Hostname.portfolio_site?(req.host) }
root 'blog#show',      constraints: ->(req) { Hostname.blog_site?(req.host) }
root 'landing#show'
```

Background: When an application has multiple `root` entries in same scope with different constraints, the current solution is to use `get '/'` instead of simply `root`. This is because when we try to add another entry of `root`, Rails complains there's already a route with `root` name.
The same doesn't happen with multiple `get`'s (or other matchers). So, I think it's fair that `root` also allows it since it's just a shortcut for a `get` internally.

PS: I couldn't find a spec related to allowing multiple `get`'s, `post`'s, etc with same path. If you know where is it, I could implement a similar spec for `root`.

rafaelsales commented:
Anything needed from my end to have feedback here?

seuros commented:
LGTM 
cc @rafaelfranca

rafaelsales commented:
ping @rafaelfranca

arthurnn commented:
Not sure about this honestly. the multiple `get '/',` look good to me already. I think this will add a few more bits of complexity to the code, that can already handle the use case described.
:-1: from me.

rafaelsales commented:
But then routes API is inconsistent. You can have multiple `root` but with a "hack", using `get '/'`
Also, where this adds complexity to code? Instead, this makes the code consistent.

rafaelfranca commented on actionpack/lib/action_dispatch/routing/mapper.rb:
```diff
@@ -586,9 +587,8 @@ def with_default_scope(scope, &block)
           end
         end
 
-        # Query if the following named route was already defined.
```
Why the documentation was removed?

rafaelsales committed:
Allow multiple `root` routes in same scope level

When an application has multiple root entries with different
constraints, the current solution is to use `get '/'`. Example:

**Currently I have to do:**
```ruby
get '/', to: 'portfolio#show', constraints: ->(req) { Hostname.portfolio_site?(req.host) }
get '/', to: 'blog#show',      constraints: ->(req) { Hostname.blog_site?(req.host) }
root 'landing#show'
```

**But I would like to do:**
```ruby
root 'portfolio#show', constraints: ->(req) { Hostname.portfolio_site?(req.host) }
root 'blog#show',      constraints: ->(req) { Hostname.blog_site?(req.host) }
root 'landing#show'
```

Other URL matchers such as `get`, `post`, etc, already allows this, so I
think it's fair that `root` also allow it since it's just a shortcut for
a `get` internally.

rafaelsales commented:
@rafaelfranca Removed that doc by mistake. Fixed and rebased

rafaelsales commented:
ty
